lib/rubysl/timeout.rb

```diff
@@ -120,13 +120,22 @@ def self.watch_channel
     end
   end
 
-  # Spin up the thread up front to avoid a thread collision problem spinning
-  # it up lazily.
-  @controller = Thread.new { watch_channel }
+  CONTROLLER_MUTEX = Mutex.new
+
+  def self.ensure_timeout_thread_running
+    unless @controller
+      CONTROLLER_MUTEX.synchronize do
+        unless @controller
+          @controller = Thread.new { watch_channel }
+        end
+      end
+    end
```

Comment:
Is this double-checked locking? That's only safe in Java if your variable is volatile, isn't it? How does it translate to Ruby?

Reply:
Shared writes are volatile with thread-safe objects.
But anyway, this `@controller` variable is never read here, it only keeps the Thread instance alive and is used as a spawned/not-spawned marker.

Reply:
This seems like a premature optimisation. I'd normally write this as a simple lock, but if you think it's ok then I'm fine to commit.

Reply:
You are right, it is premature here, I did not measure and there is anyway cross-thread interactions around, I'll remove it.